core/src/main/java/io/parsingdata/metal/data/ConcatenatedValueSource.java

```diff
@@ -16,63 +16,74 @@
 
 package io.parsingdata.metal.data;
 
+import static java.math.BigInteger.ZERO;
+
+import static io.parsingdata.metal.Trampoline.complete;
+import static io.parsingdata.metal.Trampoline.intermediate;
 import static io.parsingdata.metal.Util.checkNotNegative;
 import static io.parsingdata.metal.Util.checkNotNull;
 
 import java.math.BigInteger;
 import java.util.Objects;
 
+import io.parsingdata.metal.Trampoline;
 import io.parsingdata.metal.Util;
 import io.parsingdata.metal.expression.value.Value;
 
 public class ConcatenatedValueSource extends Source {
 
-    public final Value left;
-    public final Value right;
+    public final ImmutableList<Value> values;
+    public final BigInteger length;
 
-    public ConcatenatedValueSource(final Value left, final Value right) {
-        this.left = checkNotNull(left, "left");
-        this.right = checkNotNull(right, "right");
+    public ConcatenatedValueSource(final ImmutableList<Value> values, final BigInteger length) {
```

Comment:
Would you rather not let the length be calculated here? Does it ever make sense for length not being equal to the slice lengths of all values?

Reply:
I really want to restrict constructors to assigning fields and doing some basic checking (like `checkNotNull()` or `checkNotNegative()`, etc. So that's why the length is calculated outside the constructor. At the same time, maybe it's better suited to a factory method, that way the constructor is guarded against weird inputs.

Reply:
I ended up refactoring this: the `ConcatenatedValueSource` constructor is now private but called by a factory `create()` method. This does mean that the length is calculated inside, which means it's no longer possible to create a `ConcatenatedValueSource` with a length larger than is possible given the underlying data. It also very much cleaned up the implementations of both `FoldCat` and `Cat`.